src/c4/charconv.hpp

```diff
@@ -465,101 +746,96 @@ C4_ALWAYS_INLINE bool read_oct(csubstr s, I *C4_RESTRICT v)
+inline size_t _itoa2bufwithdigits(substr buf, size_t pos, size_t num_digits, csubstr val) noexcept
 {
     num_digits = num_digits > val.len ? num_digits - val.len : 0;
+    C4_ASSERT(num_digits + val.len <= buf.len);
     for(size_t i = 0; i < num_digits; ++i)
         _c4append('0');
-    return _itoa2buf(buf, pos, val);
+    return detail::_itoa2buf(buf, pos, val);
 }
-template<class T>
-size_t _itoadec2buf(substr buf)
+template<class I>
+C4_NO_INLINE size_t _itoadec2buf(substr buf) noexcept
 {
-    if(C4_LIKELY(buf.len > 0))
-    {
-        buf.str[0] = '-';
-        return detail::_itoa2buf(buf, 1, detail::itoa_min<sizeof(T)>::value_dec());
-    }
-    else
-    {
-        return detail::_itoa2buf({}, 1, detail::itoa_min<sizeof(T)>::value_dec());
-    }
-    C4_UNREACHABLE();
+    using digits_type = detail::charconv_digits<I>;
+    if(C4_UNLIKELY(buf.len < digits_type::maxdigits_dec))
+        return digits_type::maxdigits_dec;
+    buf.str[0] = '-';
+    return detail::_itoa2buf(buf, 1, digits_type::min_value_dec());
 }
 template<class I>
-size_t _itoa2buf(substr buf, I radix)
+C4_NO_INLINE size_t _itoa2buf(substr buf, I radix) noexcept
 {
+    using digits_type = detail::charconv_digits<I>;
     size_t pos = 0;
-    _c4append('-');
+    if(C4_LIKELY(buf.len > 0))
+        buf.str[pos++] = '-';
     switch(radix)
     {
     case I(10):
-        /*...........................*/ return _itoa2buf(buf, pos, itoa_min<sizeof(I)>::value_dec());
+        if(C4_UNLIKELY(buf.len < digits_type::maxdigits_dec))
+            return digits_type::maxdigits_dec;
+        /*.......................................*/ return _itoa2buf(buf, pos, digits_type::min_value_dec());
     case I(16):
-        _c4append('0'); _c4append('x'); return _itoa2buf(buf, pos, itoa_min<sizeof(I)>::value_hex());
+        if(C4_UNLIKELY(buf.len < digits_type::maxdigits_hex))
+            return digits_type::maxdigits_hex;
+        buf.str[pos++] = '0'; buf.str[pos++] = 'x'; return _itoa2buf(buf, pos, digits_type::min_value_hex());
     case I( 2):
-        _c4append('0'); _c4append('b'); return _itoa2buf(buf, pos, itoa_min<sizeof(I)>::value_bin());
+        if(C4_UNLIKELY(buf.len < digits_type::maxdigits_bin))
+            return digits_type::maxdigits_bin;
+        buf.str[pos++] = '0'; buf.str[pos++] = 'b'; return _itoa2buf(buf, pos, digits_type::min_value_bin());
     case I( 8):
-        _c4append('0'); _c4append('o'); return _itoa2buf(buf, pos, itoa_min<sizeof(I)>::value_oct());
+        if(C4_UNLIKELY(buf.len < digits_type::maxdigits_oct))
+            return digits_type::maxdigits_oct;
+        buf.str[pos++] = '0'; buf.str[pos++] = 'o'; return _itoa2buf(buf, pos, digits_type::min_value_oct());
     }
-    C4_ERROR("unknown radix");
+    C4_UNREACHABLE();
     return 0;
 }
 template<class I>
-size_t _itoa2buf(substr buf, I radix, size_t num_digits)
+C4_NO_INLINE size_t _itoa2buf(substr buf, I radix, size_t num_digits) noexcept
```

Comment:
Don't we have static representation of all the possible min values ?
Wouldn't it be faster to memcpy?

Reply:
Alas, we don't have the representation. That's because `num_digits` is variable and the padding zeroes need to be interleaved between the `0b` and the proper number. So we only store the actual digits, and build the padded representation on demand.

Also, it is important to note that this code path is just for the one single value per signed type: the signed min value (hence the usage of `C4_NO_INLINE`). Every other value (signed or unsigned) will go into `write_dec()`, with negative numbers being negated before that. This is required because negating the min value will overflow it to the original value: eg `int8_t(-128)=+128=127+1=-128`.

Reply:
While writing this, it occurred to me that instead of all having these dedicated functions for a single value, I could write instead `std::numeric_limits<T>::min()+1` and then just fix the last digit. Way less code to maintain!